File: packages/python/fedbpp/plan_evaluation.py
"""Deterministic v1.6 PLAN evaluation built on canonical PLAN coverage."""
from __future__ import annotations
from typing import Any

from ._analysis.coverage import analyze_plan
from ._analysis.targets import compare_to_targets
from ._analysis.policies import normalize_range, representative_scalar

EVALUATION_VERSION = "0.1.0"
EVALUATION_POLICY = "plan-evaluation-v1"

def _doc(value: Any) -> dict[str, Any]: return value.document if hasattr(value, "document") else value
def _exercise(db: Any, eid: str) -> Any:
    try: return db.get_exercise(eid) if hasattr(db, "get_exercise") else db.get("exercises", {}).get(eid)
    except KeyError: return None
def _data(ex: Any) -> dict[str, Any]: return ex.data if hasattr(ex, "data") else ex
def _range_value(value: Any) -> dict[str, float | None]:
    if not isinstance(value, dict): return {"min": None, "target": float(value) if isinstance(value, (int, float)) else None, "max": None}
    return {"min": value.get("min", value.get("minimumSets")), "target": value.get("target", value.get("targetSets")), "max": value.get("max", value.get("maximumSets"))}
def _state(actual: float, target: dict[str, float | None]) -> str:
    if target.get("min") is not None and actual < target["min"]: return "below_minimum"
    if target.get("max") is not None and actual > target["max"]: return "above_maximum"
    if target.get("target") is None: return "within_range"
    if actual == target["target"]: return "at_target"
    return "within_range_below_target" if actual < target["target"] else "within_range_above_target"
def _finding(kind: str, **values: Any) -> dict[str, Any]: return {"type": kind, **{key: values[key] for key in sorted(values)}}

def _profile_findings(plan: dict[str, Any], db: Any, profile: dict[str, Any], relationships: Any) -> tuple[dict[str, Any], dict[str, Any], list[dict[str, Any]], list[dict[str, Any]]]:
    prefs = profile.get("exercisePreferences", {}) or {}; constraints = profile.get("constraints", {}) or {}
    preferred_ex = set(prefs.get("preferredExerciseIds", [])); avoided_ex = set(prefs.get("avoidedExerciseIds", []))
    preferred_fam = set(prefs.get("preferredFamilyIds", [])); avoided_fam = set(prefs.get("avoidedFamilyIds", []))
    excluded_ex = set(constraints.get("excludedExerciseIds", [])); excluded_fam = set(constraints.get("excludedFamilyIds", []))
    used_ex = []; hard = []; soft = []; supported = []; unsupported = []; unknown = []
    available = set(profile.get("equipment", []) or [])
    body = {"body only", "bodyweight", "no equipment", "none"}
    for session in plan.get("sessions", []):
        sid = session.get("planSessionId")
        for rx in session.get("exercises", []):
            eid = rx.get("exerciseId"); used_ex.append(eid) if eid else None
            ex = _exercise(db, eid) if eid else None
            family = relationships.family_for(eid).family_id if relationships is not None and eid and relationships.family_for(eid) else None
            if eid in excluded_ex: hard.append(_finding("excluded_exercise", exerciseId=eid, prescriptionId=rx.get("prescriptionId"), sessionId=sid))
            if family in excluded_fam: hard.append(_finding("excluded_family", exerciseId=eid, familyId=family, prescriptionId=rx.get("prescriptionId"), sessionId=sid))
            if eid in preferred_ex: soft.append(_finding("preferred_exercise_used", exerciseId=eid, prescriptionId=rx.get("prescriptionId"), sessionId=sid))
            if eid in avoided_ex: soft.append(_finding("avoided_exercise_used", exerciseId=eid, prescriptionId=rx.get("prescriptionId"), sessionId=sid))
            if family in preferred_fam: soft.append(_finding("preferred_family_used", exerciseId=eid, familyId=family, prescriptionId=rx.get("prescriptionId"), sessionId=sid))
            if family in avoided_fam: soft.append(_finding("avoided_family_used", exerciseId=eid, familyId=family, prescriptionId=rx.get("prescriptionId"), sessionId=sid))
            if ex is None: unknown.append(_finding("unknown_exercise", exerciseId=eid, prescriptionId=rx.get("prescriptionId"), sessionId=sid)); continue
            equipment = _data(ex).get("source", {}).get("equipment")
            if equipment is None or str(equipment) in {"None", "other"}:
                unknown.append(_finding("unknown_equipment", exerciseId=eid, equipment=equipment, prescriptionId=rx.get("prescriptionId"), sessionId=sid))
            elif str(equipment) in body and (available & body): supported.append(eid)
            elif equipment in available: supported.append(eid)
            else: unsupported.append(_finding("unsupported_equipment", equipment=equipment, exerciseId=eid, prescriptionId=rx.get("prescriptionId"), sessionId=sid)); hard.append(unsupported[-1])
    equipment = {"supportedExercises": sorted(set(supported)), "unsupportedExercises": sorted(unsupported, key=lambda x: (x.get("exerciseId") or "", x.get("sessionId") or "", x.get("prescriptionId") or "")), "unknownEquipmentExercises": sorted(unknown, key=lambda x: (x.get("exerciseId") or "", x.get("sessionId") or "", x.get("prescriptionId") or ""))}
    preferences = {"preferredExercisesUsed": sorted({x["exerciseId"] for x in soft if x["type"] == "preferred_exercise_used"}), "preferredFamiliesUsed": sorted({x["familyId"] for x in soft if x["type"] == "preferred_family_used"}), "avoidedExercisesUsed": sorted({x["exerciseId"] for x in soft if x["type"] == "avoided_exercise_used"}), "avoidedFamiliesUsed": sorted({x["familyId"] for x in soft if x["type"] == "avoided_family_used"}), "findings": sorted(soft, key=lambda x: (x["type"], x.get("exerciseId") or x.get("familyId") or "", x.get("sessionId") or "", x.get("prescriptionId") or ""))}
    return equipment, preferences, hard, soft

def evaluate_plan(plan: Any, db: Any, profile: Any = None, target: Any = None, relationships: Any = None) -> dict[str, Any]:
    plan, db = _doc(plan), db
    profile = _doc(profile) if profile is not None else None; target = _doc(target) if target is not None else None
    coverage = analyze_plan(plan, db)
    muscle = {}
    if target is not None:
        comparison = compare_to_targets(plan, target, db)
        muscle = comparison["muscles"]
    frequency = {}
    if target is not None:
        for muscle_id, spec in sorted((target.get("frequency", {}).get("muscles", {}) or {}).items()):
            planned = coverage["exposureFrequency"]["muscles"].get(muscle_id, {"exposuresPerNativeCycle": 0, "normalizedExposuresPer7Days": 0})
            limits = _range_value(spec); actual = planned["normalizedExposuresPer7Days"]
            frequency[muscle_id] = {"plannedExposuresPerNativeCycle": planned["exposuresPerNativeCycle"], "normalizedExposuresPer7Days": actual, "minimum": limits["min"], "target": limits["target"], "maximum": limits["max"], "state": _state(actual, limits)}
    patterns = {}
    pattern_targets = target.get("movementPatterns", {}) if target else {}
    for pattern_id, spec in sorted(pattern_targets.items()):
        actual = coverage["nativeCycle"]["movementPatternSets"].get(pattern_id, 0); limits = _range_value(spec)
        patterns[pattern_id] = {"plannedSets": actual, "minimum": limits["min"], "target": limits["target"], "maximum": limits["max"], "state": _state(actual, limits)}
    families = {}
    if relationships is not None:
        from .relationships import family_coverage
        family_data = family_coverage(plan, relationships)
        for family_id, spec in sorted((target.get("families", {}) if target else {}).items()):
            actual = family_data.get(family_id, {}).get("plannedSets", 0); limits = _range_value(spec)
            families[family_id] = {"plannedSets": actual, "minimum": limits["min"], "target": limits["target"], "maximum": limits["max"], "state": _state(actual, limits)}
        family_section = {key: family_data[key] for key in sorted(family_data)}
    else: family_section = {"available": False, "reason": "relationship artifact not provided"}
    equipment = preferences = {"findings": []}; hard = []
    if profile is not None: equipment, preferences, hard, _ = _profile_findings(plan, db, profile, relationships)
    availability = {"plannedSessions": len(plan.get("sessions", []))}
    if profile is not None and profile.get("availability"):
        av = profile["availability"]; limits = av.get("sessionsPerCycle", {}) or {}; planned = len(plan.get("sessions", [])); availability.update({"min": limits.get("min"), "target": limits.get("target"), "max": limits.get("max"), "state": _state(planned, _range_value(limits))})
        excluded_days = set(av.get("excludedDayOffsets", [])); hard.extend(_finding("excluded_day_offset", sessionId=s.get("planSessionId"), dayOffset=s.get("dayOffset")) for s in plan.get("sessions", []) if s.get("dayOffset") in excluded_days)
    else: availability["state"] = "not_evaluated"
    exercise_counts = {}
    if profile is not None:
        limits = _range_value((profile.get("availability", {}) or {}).get("exercisesPerSession", {}))
        if any(value is not None for value in limits.values()):
            for session in plan.get("sessions", []):
                count = len(session.get("exercises", [])); state = _state(count, limits)
                exercise_counts[session.get("planSessionId")] = {"exerciseCount": count, "minimum": limits["min"], "target": limits["target"], "maximum": limits["max"], "state": state}
                if state in {"below_minimum", "above_maximum"}: hard.append(_finding("exercise_count", sessionId=session.get("planSessionId"), exerciseCount=count, minimum=limits["min"], maximum=limits["max"]))
                elif limits["target"] is not None and count != limits["target"]: preferences.setdefault("findings", []).append(_finding("exercise_count_target_miss", sessionId=session.get("planSessionId"), exerciseCount=count, target=limits["target"]))
    target_gaps = [x for x in muscle.values() if x.get("state") == "below_minimum"] + [x for x in frequency.values() if x.get("state") == "below_minimum"] + [x for x in patterns.values() if x.get("state") == "below_minimum"] + [x for x in families.values() if x.get("state") == "below_minimum"]
    completeness = coverage["coverageCompleteness"]
    warnings = []
    if completeness.get("unmappedSets", 0) or completeness.get("ineligibleSets", 0) or equipment.get("unknownEquipmentExercises"): warnings.append("coverage is incomplete for one or more PLAN exercises")
    if profile and profile.get("availability", {}).get("minutesPerSession"): warnings.append("duration estimation is not evaluated by duration-estimation-v1 because PLAN rest/transition inputs are not complete")
    if target and target.get("families") and relationships is None: warnings.append("family targets cannot be evaluated because relationship artifact was not provided")
    md = coverage["analysisMetadata"]; dbmd = db.metadata if hasattr(db, "metadata") else db.get("metadata", {})
    prov = {"analysisVersion": EVALUATION_VERSION, "analysisPolicy": EVALUATION_POLICY, "planSchemaVersion": plan.get("schemaVersion"), "profileSchemaVersion": profile.get("schemaVersion") if profile else None, "targetSchemaVersion": target.get("schemaVersion") if target else None, "relationshipSchemaVersion": relationships.document.get("schemaVersion") if hasattr(relationships, "document") else (relationships.get("schemaVersion") if relationships else None), "dbSchemaVersion": md.get("dbSchemaVersion"), "dbConverterVersion": md.get("dbConverterVersion"), "dbUpstreamSha256": md.get("dbUpstreamSha256"), "setCredits": md.get("setCredits"), "durationPolicy": "duration-estimation-v1" if profile and profile.get("availability", {}).get("minutesPerSession") else None}
    # Prefer the canonical provenance keys when DB++ metadata is present.
    for key in ("schemaVersion", "converterVersion"):
        if prov["db" + ("SchemaVersion" if key == "schemaVersion" else "ConverterVersion")] is None: prov["db" + ("SchemaVersion" if key == "schemaVersion" else "ConverterVersion")] = dbmd.get(key)
    if prov["dbUpstreamSha256"] is None: prov["dbUpstreamSha256"] = dbmd.get("upstream", {}).get("sha256")
    hard = sorted(hard, key=lambda x: (x["type"], x.get("exerciseId") or x.get("familyId") or x.get("sessionId") or "", x.get("sessionId") or "", x.get("prescriptionId") or ""))
    status = "hard_constraint_violation" if hard else ("incomplete_coverage" if completeness.get("mappedFraction", 1) < 1 or (target and target.get("families") and relationships is None) else ("valid_with_target_gaps" if target_gaps else "valid"))
    return {"summary": {"hardConstraintViolations": len(hard), "targetGaps": len(target_gaps), "softPreferenceWarnings": len(preferences.get("findings", [])), "satisfiesHardConstraints": not hard, "meetsTargetMinimums": not target_gaps, "evaluationStatus": status}, "muscleCoverage": muscle, "frequency": frequency, "movementPatterns": patterns, "families": {"coverage": family_section, "targets": families}, "equipment": equipment, "availability": availability, "exerciseCounts": exercise_counts, "preferences": preferences, "constraints": {"violations": hard}, "coverageCompleteness": completeness, "warnings": sorted(set(warnings)), "provenance": prov}

__all__ = ["evaluate_plan"]

File: packages/kotlin/fedbpp/fedbpp/src/main/kotlin/com/fedbpp/PlanEvaluation.kt
package com.fedbpp

import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.booleanOrNull
import kotlinx.serialization.json.buildJsonArray
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.doubleOrNull
import kotlinx.serialization.json.intOrNull
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.put
import kotlin.math.round

/**
 * Native implementation of the canonical PLAN evaluator.  Its JSON result is
 * deliberately shaped like Python's ``fedbpp.evaluate_plan`` result so the
 * generator and public engine façade share one evaluator gate.
 */
private const val EVALUATION_VERSION = "0.1.0"
private const val EVALUATION_POLICY = "plan-evaluation-v1"

private fun JsonElement?.obj(): JsonObject = this as? JsonObject ?: JsonObject(emptyMap())
private fun JsonElement?.arr(): JsonArray = this as? JsonArray ?: JsonArray(emptyList())
private fun JsonElement?.string(): String? = (this as? JsonPrimitive)?.content
private fun JsonElement?.number(): Double? = (this as? JsonPrimitive)?.doubleOrNull
private fun clean(value: Double): Double = round(value * 1_000_000.0) / 1_000_000.0
private fun jo(vararg values: Pair<String, JsonElement>): JsonObject = JsonObject(linkedMapOf(*values))
private fun jn(value: Double?): JsonElement = value?.let(::JsonPrimitive) ?: JsonNull

private data class Range(val min: Double? = null, val target: Double? = null, val max: Double? = null)
private fun range(value: JsonElement?): Range {
    if (value !is JsonObject) return Range(target = value.number())
    return Range(value["min"]?.number() ?: value["minimumSets"]?.number(), value["target"]?.number() ?: value["targetSets"]?.number(), value["max"]?.number() ?: value["maximumSets"]?.number())
}
private fun state(actual: Double, range: Range): String = when {
    range.min != null && actual < range.min -> "below_minimum"
    range.max != null && actual > range.max -> "above_maximum"
    range.target == null -> "within_range"
    actual == range.target -> "at_target"
    actual < range.target -> "within_range_below_target"
    else -> "within_range_above_target"
}
private fun plannedSets(rx: JsonObject): Double {
    val explicit = rx["plannedSets"]?.arr()
    if (explicit != null && explicit.isNotEmpty()) return explicit.size.toDouble()
    val v = rx["sets"]
    return when (v) { is JsonPrimitive -> v.doubleOrNull ?: 0.0; is JsonObject -> range(v).target ?: range(v).min ?: range(v).max ?: 0.0; else -> 0.0 }
}

private data class Coverage(
    val direct: MutableMap<String, Double> = sortedMapOf(), val indirect: MutableMap<String, Double> = sortedMapOf(),
    val stabilizer: MutableMap<String, Double> = sortedMapOf(), val patterns: MutableMap<String, Double> = sortedMapOf(),
    val muscleSessions: MutableMap<String, MutableSet<String>> = sortedMapOf(), val patternSessions: MutableMap<String, MutableSet<String>> = sortedMapOf(),
    var planned: Double = 0.0, var mapped: Double = 0.0, var unmapped: Double = 0.0, var ineligible: Double = 0.0,
    val unmappedRx: MutableList<String> = mutableListOf(), val ineligibleRx: MutableList<String> = mutableListOf()
)

private fun add(table: MutableMap<String, Double>, key: String, value: Double) { table[key] = clean((table[key] ?: 0.0) + value) }
private fun mapNumbers(input: Map<String, Double>): JsonObject = buildJsonObject { input.toSortedMap().forEach { (k, v) -> put(k, JsonPrimitive(clean(v))) } }
private fun setCredits(database: Database): Triple<Double, Double, Double> {
    val credits = database.metadata["setCredits"].obj()
    return Triple(credits["direct"].number() ?: 1.0, credits["indirect"].number() ?: 0.5, credits["stabilizer"].number() ?: 0.0)
}

private fun coverage(plan: JsonObject, database: Database): Coverage {
    val result = Coverage()
    for (sessionValue in plan["sessions"].arr()) {
        val session = sessionValue.obj(); val sid = session["planSessionId"].string() ?: ""
        for (rxValue in session["exercises"].arr()) {
            val rx = rxValue.obj(); val sets = plannedSets(rx); result.planned += sets
            val id = rx["exerciseId"].string(); val exercise = id?.let { database.exercises[it] }
            if (exercise == null) { result.unmapped += sets; rx["prescriptionId"].string()?.let(result.unmappedRx::add); continue }
            result.mapped += sets
            if (!exercise.annotation.volumeEligible) { result.ineligible += sets; rx["prescriptionId"].string()?.let(result.ineligibleRx::add); continue }
            exercise.annotation.direct.forEach { add(result.direct, it, sets); result.muscleSessions.getOrPut(it) { sortedSetOf() }.add(sid) }
            exercise.annotation.indirect.forEach { add(result.indirect, it, sets); result.muscleSessions.getOrPut(it) { sortedSetOf() }.add(sid) }
            exercise.annotation.stabilizers.forEach { add(result.stabilizer, it, sets) }
            exercise.annotation.patterns.forEach { add(result.patterns, it, sets); result.patternSessions.getOrPut(it) { sortedSetOf() }.add(sid) }
        }
    }
    return result
}

private fun finding(type: String, vararg values: Pair<String, JsonElement>): JsonObject = buildJsonObject { put("type", JsonPrimitive(type)); values.sortedBy { it.first }.forEach { put(it.first, it.second) } }

/** Evaluate a PLAN without a Python bridge or source-tree dependency. */
fun evaluatePlan(planValue: JsonElement, database: Database, profileValue: JsonElement? = null, targetValue: JsonElement? = null, relationships: ExerciseRelationships? = null): JsonElement {
    val plan = planValue.obj(); val profile = profileValue as? JsonObject; val target = targetValue as? JsonObject
    val c = coverage(plan, database); val days = plan["cycle"].obj()["lengthDays"]?.jsonPrimitive?.intOrNull ?: 7
    val credits = setCredits(database); val muscles = (c.direct.keys + c.indirect.keys + c.stabilizer.keys).toSortedSet()
    val effective = muscles.associateWith { clean((c.direct[it] ?: 0.0) * credits.first + (c.indirect[it] ?: 0.0) * credits.second + (c.stabilizer[it] ?: 0.0) * credits.third) }
    val targetDays = target?.get("periodDays")?.number() ?: days.toDouble(); val targetScale = targetDays / days
    val muscleRows = buildJsonObject {
        val configured = target?.get("muscles").obj()
        (configured.keys + effective.keys).toSortedSet().forEach { muscle ->
            val r = range(configured[muscle]); val actual = clean((effective[muscle] ?: 0.0) * targetScale)
            val row = jo(
                "actualEffectiveSets" to JsonPrimitive(actual), "minimum" to jn(r.min), "target" to jn(r.target),
                "maximum" to jn(r.max), "min" to jn(r.min), "max" to jn(r.max),
                "differenceFromTarget" to jn(r.target?.let { clean(actual - it) }),
                "planEffectiveSetRange" to jo("min" to JsonPrimitive(actual), "target" to JsonPrimitive(actual), "max" to JsonPrimitive(actual)),
                "state" to JsonPrimitive(if (configured[muscle] == null) "not_targeted" else state(actual, r)),
                "periodDays" to JsonPrimitive(targetDays)
            )
            put(muscle, row)
        }
    }
    val frequencyRows = buildJsonObject {
        target?.get("frequency").obj()["muscles"].obj().toSortedMap().forEach { (muscle, spec) ->
            val exposures = c.muscleSessions[muscle]?.size ?: 0; val normalized = clean(exposures * 7.0 / days); val r = range(spec)
            put(muscle, jo("plannedExposuresPerNativeCycle" to JsonPrimitive(exposures), "normalizedExposuresPer7Days" to JsonPrimitive(normalized), "minimum" to jn(r.min), "target" to jn(r.target), "maximum" to jn(r.max), "state" to JsonPrimitive(state(normalized, r))))
        }
    }
    val patternRows = buildJsonObject {
        target?.get("movementPatterns").obj().toSortedMap().forEach { (pattern, spec) ->
            val actual = c.patterns[pattern] ?: 0.0; val r = range(spec)
            put(pattern, jo("plannedSets" to JsonPrimitive(actual), "minimum" to jn(r.min), "target" to jn(r.target), "maximum" to jn(r.max), "state" to JsonPrimitive(state(actual, r))))
        }
    }
    val familyCounts = sortedMapOf<String, Double>(); val familyCoverage = buildJsonObject {
        if (relationships != null) database.exercises.keys.sorted().forEach { id -> relationships.familyFor(id)?.familyId?.let { family -> familyCounts.putIfAbsent(family, 0.0) } }
        for (sessionValue in plan["sessions"].arr()) for (rxValue in sessionValue.obj()["exercises"].arr()) {
            val id = rxValue.obj()["exerciseId"].string() ?: continue; relationships?.familyFor(id)?.familyId?.let { add(familyCounts, it, plannedSets(rxValue.obj())) }
        }
        familyCounts.forEach { (id, sets) -> put(id, jo("familyId" to JsonPrimitive(id), "plannedSets" to JsonPrimitive(sets))) }
    }
    val familyTargets = buildJsonObject {
        target?.get("families").obj().toSortedMap().forEach { (family, spec) ->
            val actual = familyCounts[family] ?: 0.0; val r = range(spec)
            put(family, jo("plannedSets" to JsonPrimitive(actual), "minimum" to jn(r.min), "target" to jn(r.target), "maximum" to jn(r.max), "state" to JsonPrimitive(state(actual, r))))
        }
    }
    val hard = mutableListOf<JsonObject>(); val soft = mutableListOf<JsonObject>(); val supported = sortedSetOf<String>(); val unsupported = mutableListOf<JsonObject>(); val unknown = mutableListOf<JsonObject>()
    val availability = profile?.get("availability").obj(); val excludedDays = availability["excludedDayOffsets"].arr().mapNotNull { it.jsonPrimitive.intOrNull }.toSet()
    val preferences = profile?.get("exercisePreferences").obj(); val constraints = profile?.get("constraints").obj()
    val excludedExercise = constraints["excludedExerciseIds"].arr().mapNotNull { it.string() }.toSet(); val excludedFamilies = constraints["excludedFamilyIds"].arr().mapNotNull { it.string() }.toSet()
    val preferredExercise = preferences["preferredExerciseIds"].arr().mapNotNull { it.string() }.toSet(); val avoidedExercise = preferences["avoidedExerciseIds"].arr().mapNotNull { it.string() }.toSet(); val preferredFamilies = preferences["preferredFamilyIds"].arr().mapNotNull { it.string() }.toSet(); val avoidedFamilies = preferences["avoidedFamilyIds"].arr().mapNotNull { it.string() }.toSet()
    val availableEquipment = profile?.get("equipment").arr().mapNotNull { it.string() }.toSet(); val body = setOf("body only", "bodyweight", "no equipment", "none")
    for (sessionValue in plan["sessions"].arr()) {
        val session = sessionValue.obj(); val sid = session["planSessionId"].string() ?: ""
        if (session["dayOffset"]?.jsonPrimitive?.intOrNull in excludedDays) hard += finding("excluded_day_offset", "sessionId" to JsonPrimitive(sid), "dayOffset" to (session["dayOffset"] ?: JsonNull))
        for (rxValue in session["exercises"].arr()) {
            val rx = rxValue.obj(); val id = rx["exerciseId"].string() ?: continue; val pid = rx["prescriptionId"].string() ?: ""
            val family = relationships?.familyFor(id)?.familyId
            if (id in excludedExercise) hard += finding("excluded_exercise", "exerciseId" to JsonPrimitive(id), "prescriptionId" to JsonPrimitive(pid), "sessionId" to JsonPrimitive(sid))
            if (family in excludedFamilies) hard += finding("excluded_family", "exerciseId" to JsonPrimitive(id), "familyId" to JsonPrimitive(family!!), "prescriptionId" to JsonPrimitive(pid), "sessionId" to JsonPrimitive(sid))
            if (id in preferredExercise) soft += finding("preferred_exercise_used", "exerciseId" to JsonPrimitive(id), "prescriptionId" to JsonPrimitive(pid), "sessionId" to JsonPrimitive(sid))
            if (id in avoidedExercise) soft += finding("avoided_exercise_used", "exerciseId" to JsonPrimitive(id), "prescriptionId" to JsonPrimitive(pid), "sessionId" to JsonPrimitive(sid))
            if (family in preferredFamilies) soft += finding("preferred_family_used", "exerciseId" to JsonPrimitive(id), "familyId" to JsonPrimitive(family!!), "prescriptionId" to JsonPrimitive(pid), "sessionId" to JsonPrimitive(sid))
            if (family in avoidedFamilies) soft += finding("avoided_family_used", "exerciseId" to JsonPrimitive(id), "familyId" to JsonPrimitive(family!!), "prescriptionId" to JsonPrimitive(pid), "sessionId" to JsonPrimitive(sid))
            val exercise = database.exercises[id]
            if (exercise == null) { unknown += finding("unknown_exercise", "exerciseId" to JsonPrimitive(id), "prescriptionId" to JsonPrimitive(pid), "sessionId" to JsonPrimitive(sid)); continue }
            val equipment = exercise.source["equipment"].string()
            when { equipment == null || equipment in setOf("None", "other") -> unknown += finding("unknown_equipment", "exerciseId" to JsonPrimitive(id), "equipment" to (exercise.source["equipment"] ?: JsonNull), "prescriptionId" to JsonPrimitive(pid), "sessionId" to JsonPrimitive(sid))
                equipment in body && availableEquipment.intersect(body).isNotEmpty() -> supported += id
                equipment in availableEquipment -> supported += id
                else -> { val f = finding("unsupported_equipment", "equipment" to JsonPrimitive(equipment), "exerciseId" to JsonPrimitive(id), "prescriptionId" to JsonPrimitive(pid), "sessionId" to JsonPrimitive(sid)); unsupported += f; hard += f }
            }
        }
    }
    val exerciseCounts = buildJsonObject {
        val r = range(availability["exercisesPerSession"]); if (profile != null && listOf(r.min, r.target, r.max).any { it != null }) for (sessionValue in plan["sessions"].arr()) {
            val session = sessionValue.obj(); val n = session["exercises"].arr().size.toDouble(); val s = state(n, r); val sid = session["planSessionId"].string() ?: ""
            put(sid, jo("exerciseCount" to JsonPrimitive(n.toInt()), "minimum" to jn(r.min), "target" to jn(r.target), "maximum" to jn(r.max), "state" to JsonPrimitive(s)))
            if (s == "below_minimum" || s == "above_maximum") hard += finding("exercise_count", "sessionId" to JsonPrimitive(sid), "exerciseCount" to JsonPrimitive(n.toInt()), "minimum" to jn(r.min), "maximum" to jn(r.max))
            else if (r.target != null && n != r.target) soft += finding("exercise_count_target_miss", "sessionId" to JsonPrimitive(sid), "exerciseCount" to JsonPrimitive(n.toInt()), "target" to jn(r.target))
        }
    }
    val sessionRange = range(availability["sessionsPerCycle"]); val sessionCount = plan["sessions"].arr().size.toDouble(); val availabilityResult = if (profile == null || availability.isEmpty()) jo("plannedSessions" to JsonPrimitive(sessionCount.toInt()), "state" to JsonPrimitive("not_evaluated")) else jo("plannedSessions" to JsonPrimitive(sessionCount.toInt()), "min" to jn(sessionRange.min), "target" to jn(sessionRange.target), "max" to jn(sessionRange.max), "state" to JsonPrimitive(state(sessionCount, sessionRange)))
    val targetGaps = listOf(muscleRows, frequencyRows, patternRows, familyTargets).sumOf { rows -> rows.values.count { it.obj()["state"].string() == "below_minimum" } }
    val mappedFraction = if (c.planned == 0.0) 1.0 else clean(c.mapped / c.planned)
    val completeness = jo("plannedSets" to JsonPrimitive(clean(c.planned)), "plannedSetRange" to jo("min" to JsonPrimitive(clean(c.planned)), "target" to JsonPrimitive(clean(c.planned)), "max" to JsonPrimitive(clean(c.planned))), "mappedSets" to JsonPrimitive(clean(c.mapped)), "mappedSetRange" to jo("min" to JsonPrimitive(clean(c.mapped)), "target" to JsonPrimitive(clean(c.mapped)), "max" to JsonPrimitive(clean(c.mapped))), "unmappedSets" to JsonPrimitive(clean(c.unmapped)), "unmappedSetRange" to jo("min" to JsonPrimitive(clean(c.unmapped)), "target" to JsonPrimitive(clean(c.unmapped)), "max" to JsonPrimitive(clean(c.unmapped))), "ineligibleSets" to JsonPrimitive(clean(c.ineligible)), "ineligibleSetRange" to jo("min" to JsonPrimitive(clean(c.ineligible)), "target" to JsonPrimitive(clean(c.ineligible)), "max" to JsonPrimitive(clean(c.ineligible))), "mappedFraction" to JsonPrimitive(mappedFraction), "unmappedPrescriptions" to JsonArray(c.unmappedRx.sorted().map(::JsonPrimitive)), "ineligiblePrescriptions" to JsonArray(c.ineligibleRx.sorted().map(::JsonPrimitive)))
    val warnings = mutableSetOf<String>(); if (c.unmapped != 0.0 || c.ineligible != 0.0 || unknown.isNotEmpty()) warnings += "coverage is incomplete for one or more PLAN exercises"; if (profile?.get("availability").obj()["minutesPerSession"] != null) warnings += "duration estimation is not evaluated by duration-estimation-v1 because PLAN rest/transition inputs are not complete"; if (target?.get("families").obj()?.isNotEmpty() == true && relationships == null) warnings += "family targets cannot be evaluated because relationship artifact was not provided"
    val sortedHard = hard.sortedBy { "${it["type"].string()}:${it["exerciseId"].string() ?: it["familyId"].string() ?: it["sessionId"].string() ?: ""}:${it["sessionId"].string() ?: ""}:${it["prescriptionId"].string() ?: ""}" }
    val status = if (hard.isNotEmpty()) "hard_constraint_violation" else if (mappedFraction < 1 || (target?.get("families").obj()?.isNotEmpty() == true && relationships == null)) "incomplete_coverage" else if (targetGaps > 0) "valid_with_target_gaps" else "valid"
    return buildJsonObject {
        put("summary", jo("hardConstraintViolations" to JsonPrimitive(sortedHard.size), "targetGaps" to JsonPrimitive(targetGaps), "softPreferenceWarnings" to JsonPrimitive(soft.size), "satisfiesHardConstraints" to JsonPrimitive(sortedHard.isEmpty()), "meetsTargetMinimums" to JsonPrimitive(targetGaps == 0), "evaluationStatus" to JsonPrimitive(status)))
        put("muscleCoverage", muscleRows); put("frequency", frequencyRows); put("movementPatterns", patternRows); put("families", jo("coverage" to if (relationships == null) jo("available" to JsonPrimitive(false), "reason" to JsonPrimitive("relationship artifact not provided")) else familyCoverage, "targets" to familyTargets)); put("equipment", jo("supportedExercises" to JsonArray(supported.map(::JsonPrimitive)), "unsupportedExercises" to JsonArray(unsupported.sortedBy { it["exerciseId"].string() ?: "" }), "unknownEquipmentExercises" to JsonArray(unknown.sortedBy { it["exerciseId"].string() ?: "" }))); put("availability", availabilityResult); put("exerciseCounts", exerciseCounts)
        put("preferences", jo("preferredExercisesUsed" to JsonArray(soft.filter { it["type"].string() == "preferred_exercise_used" }.mapNotNull { it["exerciseId"] }.distinct().sortedBy { it.string() }), "preferredFamiliesUsed" to JsonArray(soft.filter { it["type"].string() == "preferred_family_used" }.mapNotNull { it["familyId"] }.distinct().sortedBy { it.string() }), "avoidedExercisesUsed" to JsonArray(soft.filter { it["type"].string() == "avoided_exercise_used" }.mapNotNull { it["exerciseId"] }.distinct().sortedBy { it.string() }), "avoidedFamiliesUsed" to JsonArray(soft.filter { it["type"].string() == "avoided_family_used" }.mapNotNull { it["familyId"] }.distinct().sortedBy { it.string() }), "findings" to JsonArray(soft.sortedBy { "${it["type"].string()}:${it["exerciseId"].string() ?: it["familyId"].string() ?: ""}" })))
        put("constraints", jo("violations" to JsonArray(sortedHard))); put("coverageCompleteness", completeness); put("warnings", JsonArray(warnings.sorted().map(::JsonPrimitive)))
        put("provenance", jo("analysisVersion" to JsonPrimitive(EVALUATION_VERSION), "analysisPolicy" to JsonPrimitive(EVALUATION_POLICY), "planSchemaVersion" to (plan["schemaVersion"] ?: JsonNull), "profileSchemaVersion" to (profile?.get("schemaVersion") ?: JsonNull), "targetSchemaVersion" to (target?.get("schemaVersion") ?: JsonNull), "relationshipSchemaVersion" to (relationships?.schemaVersion?.let(::JsonPrimitive) ?: JsonNull), "dbSchemaVersion" to (database.metadata["schemaVersion"] ?: JsonNull), "dbConverterVersion" to (database.metadata["converterVersion"] ?: JsonNull), "dbUpstreamSha256" to (database.metadata["upstream"].obj()["sha256"] ?: JsonNull), "setCredits" to jo("direct" to JsonPrimitive(credits.first), "indirect" to JsonPrimitive(credits.second), "stabilizer" to JsonPrimitive(credits.third)), "durationPolicy" to if (profile?.get("availability").obj()["minutesPerSession"] != null) JsonPrimitive("duration-estimation-v1") else JsonNull))
    }
}
